src/Shape.ts

```diff
@@ -1,3 +1,45 @@
+import { Point } from './Point';
+import { ShapeException } from './ShapeException';
+
 export abstract class Shape {
+  protected color: string;
+  protected filled: boolean;
+  protected points: Point[];
+
+  constructor(points: Point[]);
+  constructor(points: Point[], color: string, filled: boolean);
+  constructor(points: Point[], color?: string, filled?: boolean) {
+    this.color = color ?? 'green';
+    this.filled = filled ?? true;
+    this.points = points;
+    this.validateShape();
```

Comment:
If you move the validation before the main code, it will return the error earlier, which is always better.